zilliqa/src/block_store.rs

```diff
@@ -112,57 +555,91 @@ impl BlockStore {
             db: self.db.clone(),
             block_cache: self.block_cache.clone(),
             highest_known_view: 0,
-            peers: HashMap::new(),
+            highest_confirmed_view: 0,
+            peers: BTreeMap::new(),
             requested_view: 0,
             max_blocks_in_flight: 0,
-            batch_size: 0,
             failed_request_sleep_duration: Duration::ZERO,
-            buffered: LruCache::new(NonZeroUsize::new(1).unwrap()),
-            unserviceable_requests: BTreeSet::new(),
+            strategies: self.strategies.clone(),
+            available_blocks: Vec::new(),
+            available_blocks_updated: None,
+            buffered: BlockCache::new(0),
+            unserviceable_requests: None,
             message_sender: self.message_sender.clone(),
+            clock: 0,
         })
     }
 
+    /// Update someone else's availability
+    pub fn update_availability(
+        &mut self,
+        from: PeerId,
+        avail: &Option<Vec<BlockStrategy>>,
+    ) -> Result<()> {
+        let the_peer = self.peer_info(from);
+        the_peer.availability.strategies.clone_from(avail);
+        the_peer.availability_updated_at = Some(SystemTime::now());
+        Ok(())
+    }
+
+    /// Retrieve our availability.
+    /// We need to do this by view range, which means that we need to account for views where there was no block.
+    /// So, the underlying db function finds the view lower and upper bounds of our contiguous block ranges and we
+    /// advertise those.
+    /// This function overlays that with a timeout so we don't ask (and thus load the db) too often.
+    pub fn availability(&mut self) -> Result<Option<Vec<BlockStrategy>>> {
+        let mut to_return = self.strategies.clone();
+        let now = SystemTime::now();
+        if self.available_blocks_updated.map_or(true, |x| {
+            now.duration_since(x).unwrap_or(Duration::ZERO)
+                > Duration::from_secs(constants::RECOMPUTE_BLOCK_AVAILABILITY_AFTER_S)
```

Comment:
yes, absolutely - but this is a fairly expensive db query, so timer triggered to avoid doing it too often and causing too great a load; this way, it gets done eventually, but we bound the amount of work we are prepared to do to keep it up to date. Smarter algorithms certainly possible, just ... harder :-)

Reply:
(in particular, guards against block proposal floods stalling the node by causing it to continually recompute availability; could as easily as you say keep this in memory and recompute this on block insert rather than availability query - however, that would've been a bit more intrusive, so I didn't)